docs/ApiGuidelines.md

```diff
@@ -0,0 +1,76 @@
+* Public APIs cannot have source-breaking changes across Open Enclave versions. For platforms 
+  where shared libraries are build, public APIs cannot have binary-breaking changes across 
+  Open Enclave versions either. (Source- and binary-breaking changes can still be made to a new 
+  API before it is released for the first time.)
+* A public API must not be added to Open Enclave if there is already a non-deprecated API that can just 
+  as easily be used (i.e., with approximately the same number of lines of code).
+* A public API must not be added to Open Enclave if the API would provide generic functionality
+  (i.e., not Open Enclave specific functionality) that could be provided by an external library.
+* All new Public C API names must start with oe_.
+* All public headers must follow Open Enclave [coding conventions](DevelopmentGuide.md).
+
+### Experimental APIs
+
+Experimental APIs are considered to be potential future public APIs, but there is no guarantee 
+of support in future releases, nor is there any guarantee that breaking changes will not occur 
+across releases. 
+
+Experimental API requirements: 
+
+* Experimental APIs must be optional to compile in. That is, there must be some way to build
+  Open Enclave in a way that does not expose any experimental APIs, so that an application
+  can easily verify that it uses only fully-supported public APIs, and so that the Open Enclave
+  code size requirement can be minimized.
+* Experimental APIs must be in a header file that is published to the build directory. They
+  can be in their own header, or in a header file that is shared with public APIs but only
+  if surrounded by an appropriate ifdef.
+* Experimental APIs should be fully documented using doxygen markup in the header file
+* Experimental APIs should ideally be used in some sample code (separate from samples that
+  use only public APIs) or snippet. Ideally, the sample should be one that is compiled as part
+  of the build. For rarely used APIs, the sample might simply be a snippet in the doxygen
+  comments. Having working samples may help others evaluate and determine how ready it is for
+  prime time before the API moves from experimental to public.
+
+### Internal APIs
+
+Internal APIs are considered to be usable within Open Enclave but are not intended to be
+used by applications that use Open Enclave. 
+
+Internal API requirements: 
+
+* Internal APIs must not appear in any header file that is published to the build directory
+* Internal APIs should ideally be documented using doxygen markup in the private header files
+* Internal APIs must not be used in sample code
+
+## Breaking Changes
+
+* Public APIs cannot have source-breaking changes (and, for platforms where shared libraries or built,
+  cannot have binary-breaking changes) across releases, except as covered by deprecation as explained
+  below. API additions can be made at any time.
```

Comment:
Do we have a stated amount of time (or number of releases) that we will take to deprecate and then remove a public API? Of course, if a security issue appears with one of our public APIs we'll deprecate it and remove it immediately. Maybe we should we leave this unstated and leave it up to the committers to decide.

Reply:
Can you deprecate in a minor release and then replace in the next major release?

Reply:
https://semver.org/ just says you have to bump the MAJOR version number to remove a public API, which seems fine to me.

Reply:
CGC discussion: when at 0.* then can break changes on MINOR version number, if deprecated in previous MINOR version number.